app/src/main/java/com/jshvarts/todoapp/arch/MviViewModel.kt

```diff
@@ -11,18 +16,28 @@ import androidx.lifecycle.ViewModel
  * ```
  */
 abstract class MviViewModel<A : UiAction, S : UiState, E : UiEffect>(
-  private val savedStateHandle: SavedStateHandle? = null
-) : ViewModel(), ActionDispatcher<A> {
-  abstract val initialState: S
-  abstract val savedStateHandleKey: String?
-  abstract val actionHandler: (A) -> Unit
-
-  private val isStateInBundle: Boolean
-    get() = if (savedStateHandleKey != null && savedStateHandle != null) {
-      initialState != savedStateHandle[savedStateHandleKey!!]
-    } else false
-
-  override fun dispatchAction(action: A) {
-    return if (isStateInBundle) Unit else actionHandler.invoke(action)
-  }
-}
\ No newline at end of file
+    private val savedStateHandle: SavedStateHandle? = null
+) : ViewModel() {
+    protected abstract val initialState: S
+
+    private val _uiEffect = MutableSharedFlow<E>()
+    val uiEffect: SharedFlow<E> = _uiEffect.asSharedFlow()
```

Comment:
This was previously in the derived classes. Seems like we could avoid duplicating the implementation.

Reply:
> This was previously in the derived classes. Seems like we could avoid duplicating the implementation.

So there is a way to populate `SharedFlow` using `Flow.shareIn` in which case you don't need a backing property `_uiEffect`. Hence it was in derived classes for flexibility. Also, not every derived view model needs to expose effects

Reply:
Kept the interface and removed the implementation.

Reply:
What happens when a derived view model does not expose any effects?

Reply:
I was wondering the same thing. Right now, we create a `SharedFlow` that emits `Nothing`. Kinda weird but not sure of a better alternative.

Reply:
just to ensure consistent naming we could:

```
interface HasUiEffect<E : UiEffect> {
  val uiEffect: SharedFlow<E>
}

// derived view model implements it (base view model does not declare it)

class NoteListViewModel @Inject constructor(
  private val savedStateHandle: SavedStateHandle,
  private val noteRepository: NoteRepository
) : MviViewModel<NoteListUiAction, NoteListUiState, NoteListUiEffect>(savedStateHandle),
  HasUiEffect<NoteListUiEffect> {

  private val _uiEffect = MutableSharedFlow<NoteListUiEffect>()
  override val uiEffect: SharedFlow<NoteListUiEffect> = _uiEffect.asSharedFlow()
```

The thing I don't like about it is now you can end up with `MviViewModel<NoteListUiAction, NoteListUiState, Nothing>(), HasUiEffect<NoteListUiEffect>` which does not make sense